Add support for term meta data.

stephenharris opened the pull request:
WordPress 4.4 adds term meta data, and [trac ticket 34602](https://core.trac.wordpress.org/ticket/34062), adds term meta data to the WXR export file. Although that ticket is still open, and is unlikely to be in 4.4, this PR adds support for importing term meta data, following the schema set out in that ticket.

stephenharris committed:
Add support for term meta data. See https://core.trac.wordpress.org/ticket/34062

stephenharris committed:
Add support for term meta data. See https://core.trac.wordpress.org/ticket/34062

stephenharris committed:
Fix docbloc return value description

stephenharris committed:
Fix docbloc return value description

Commit:
add(vcs) Add basic ignore file, see #20

Commit:
add(vcs) Add basic ignore file, see #20

Commit:
feat(pm) Add support for Composer, see #20

Commit:
feat(pm) Add support info, see #20

anhskohbo commented:
+1

rmccue commented:
Thanks for the PR, and apologies for the delayed response! Going to hold off on this until the core one is merged. :)

rmccue committed:
First pass on admin UI

rmccue committed:
Use "wordpress" as the slug

rmccue committed:
Use correct author element

rmccue committed:
Remove testing code

rmccue committed:
Add better prelim method

rmccue committed:
Add the ability to get more data about an import

rmccue committed:
Add even more info to the preliminary scan

rmccue committed:
Add summary panel to step 2

rmccue committed:
Add ability to select existing import file

rmccue committed:
Remove unused method

rmccue committed:
Cache preliminary data on the media object

rmccue committed:
Add SSE support

rmccue committed:
Add import UI

rmccue committed:
Add missing actions for completeness

rmccue committed:
Switch method to POST

rmccue committed:
Correctly set fetch_attachments

rmccue committed:
Update progress on failure too

rmccue committed:
Add default_author option

This ensures we don't try to incorrectly map authors.

rmccue committed:
Only show privately-published files in picker

wp_handle_import_upload() sets status to private, which is one of the
only things we can query by. We could also query by
_wp_attachment_context, but the JS won't let us send that.

rmccue committed:
Rework intro for the 21st Century

rmccue committed:
Remove HTML upload notice

This is added as an action, but refers to media-new.php directly, so
doesn't make sense outside of that.

rmccue committed:
Update the README

rmccue committed:
Remove unused templates and vars

rmccue committed:
Ensure upload `init` is scoped correctly

rmccue committed:
Escape HTML in error message

rmccue committed:
Add missing braces

rmccue committed:
Add docs to undocumented methods

rmccue committed:
Add missing braces

rmccue committed:
Move header and footer to functions

rmccue committed:
Handle the admin header ourselves

rmccue committed:
Move CSS and JS to external files

rmccue committed:
Pass script data to script, duh

rmccue committed:
Remove duplicated script

rmccue committed:
Add missing translation

rmccue committed:
Ensure all the text is translated and escaped

rmccue committed:
Move intro CSS to file

rmccue committed:
Use wp_kses rather than esc_html

rmccue committed:
Remove useless Dashicon

rmccue committed:
Merge pull request #26 from humanmade/admin-ui

Add admin UI

livinglab commented:
https://core.trac.wordpress.org/changeset/37240

boonebgorges commented:
https://core.trac.wordpress.org/changeset/37240

cbergen committed:
fix check for existing comment

Fixes two small typos which prevented existing comments from getting caught.

cbergen committed:
Even if post already exists, check for new comments

bobbingwide committed:
Fixes #40 NaN% for total 0

bobbingwide committed:
Fixes #40 NaN% for total 0

capuderg committed:
Post terms set with incorrect parameter type

I'm developing a plugin which is using this new WP importer and I came across a small issue in the code... The issue is that when I make the import with multiple AJAX calls the menu items import under new menus, which use the IDs of the menus that they are supposed to import into:
<img width="298" alt="menu-items-import-error" src="https://cloud.githubusercontent.com/assets/9714134/15097251/0341c2ae-1516-11e6-8911-eb2bc1b549f0.png">

The fix:
You are using the WP function `wp_set_post_terms`, in which the second parameter is [not passed correctly](https://codex.wordpress.org/Function_Reference/wp_set_post_terms#Notes). The id should be an integer, but a string is passed instead. Simply type casting it to `int`, solves the issue.

bobbingwide committed:
Fixes #40 - cater for total 0

swissspidy commented:
Pending a final review this looks good to me.

franz-josef-kaiser committed:
bug(pm) Remove email from list of support routes, see #20

franz-josef-kaiser committed:
bug(pm) Fix name and desc, see #20

franz-josef-kaiser committed:
feat(pm) Add keywords to make plugin easier to find, see #20

franz-josef-kaiser committed:
feat(pm) Add link to GH issues as support route, see #20

franz-josef-kaiser committed:
refactor(pm) Change authors from single person to link to contributors, see #20

franz-josef-kaiser committed:
fix(vcs) Add PM vendor folder to ignore list, see #20

Remove OS and IDE dependent files from local ignore list

franz-josef-kaiser committed:
refactor(pm) Change license from MIT to GPLv2, see #20

rmccue committed:
Merge pull request #21 from franz-josef-kaiser/issue-20-composer

Add support for Composer

bobbingwide committed:
oops - see http://herbmiller.me/2014/07/12/problem-solving-top-10-questions-ask-something-doesnt-work/ verses 4 and 5

gxclarke committed:
Old and new importer plugins can't both be activated since they now have the same slug name.

The slug name was changed from "wordpress-v2" to just "wordpress" in 887fd5ba0135e70228c6b6406f86d9259546b6df. As a result, having both old and new plugins activated causes the new version to not appear and the old version to not work.

rmccue committed:
Merge pull request #48 from bobbingwide/master

Fix invalid number of items to progress

rmccue committed:
Merge pull request #44 from cbergen/patch-1

fix check for existing comment

rmccue committed:
Merge pull request #45 from cbergen/patch-2

Even if post already exists, check for new comments

rmccue committed:
Merge pull request #51 from capuderg/patch-1

Post terms set with incorrect parameter type

rmccue committed:
Merge pull request #58 from gxclarke/patch-1

Old and new importer plugins can't both be activated since they now h…

rmccue committed:
Add testing harness for Travis

rmccue committed:
Add PHPUnit framework

rmccue committed:
Name Travis configuration file correctly

rmccue committed:
Add CodeSniffer ruleset

rmccue committed:
Pass arguments to PHPUnit

rmccue committed:
Use subtask for phpunit Grunt task

rmccue committed:
Ensure phpcs uses wpcs paths

rmccue committed:
Add basic activation test

rmccue committed:
Only build master

PRs are handled separately.

rmccue committed:
Remove HHVM from allowed failure list

rmccue committed:
Merge pull request #60 from humanmade/add-test-harness

Add test harness

rmccue committed:
Fix a bunch of coding standards violations

rmccue committed:
Exclude the ridiculous Yoda condition rule

rmccue committed:
Ignore uncached DB access errors

rmccue committed:
Fix more standards errors

rmccue committed:
Try correcting the name

rmccue committed:
Ban assignment in control structures

rmccue committed:
Switch to working assignment rule

rmccue committed:
Set installed_paths in config

rmccue committed:
Ensure vendor and node_modules are excluded

rmccue committed:
Fix inline code formatting

rmccue committed:
Ensure data is correctly unslashed

rmccue committed:
Remove assignment in conditional statements

rmccue committed:
Add note about disabled nonce check

rmccue committed:
Remove error suppression

rmccue committed:
Add stronger nonce check when importing

kraftner committed:
Wrong variable name

This commented out code is taken from the old Importer version where `$remote_url` was called `$url`

rmccue committed:
Merge pull request #65 from kraftner/patch-1

Wrong variable name

swissspidy committed:
Merge pull request #61 from humanmade/follow-coding-standards

Follow WordPress PHP coding standards

shivapoudel commented:
@rmccue Isn't the core one is merged till date :shipit:

joehoyle committed:
Icrement progress delta even when attachments are skipped

joehoyle committed:
If posts / terms / comments already exist, fire actions

This allows us to send delta updates to the front end for progress
updates.

roborourke committed:
main fix here is making sure only xml files can be selected when uploading

roborourke committed:
remove zip files as an option, bigger feature

roborourke committed:
code style

raphaelkross committed:
Adding upload support for .xml mime type

raphaelkross committed:
Add default WP notice style to import messages

raphaelkross committed:
Quick CSS interface adjusts

raphaelkross committed:
Fix phpcs error

raphaelkross committed:
Fix typo and add public access to method

raphaelkross committed:
Fix indentation

rmccue committed:
Merge pull request #83 from raphaelkross/master

Adds .xml to allowed mime types & improves "Import Completed" notice appearance.

roborourke committed:
Merge branch 'master' of github.com:humanmade/WordPress-Importer into minor-bugfixes

roborourke committed:
remove out of scope fix

rmccue committed:
Merge pull request #81 from humanmade/minor-bugfixes

Enable XML files to be selected in files dialogue

joehoyle committed:
Fix creating users instead of mapping.

Currently the code is doing an `isset` on the `$_POST` to check if the
user was mapped, however this value is always `isset` because the HTML
form field is always present, it's just set to "0" ("- Select -").

JJJ commented:
This still looks good!

rmccue committed:
Check for errors returned from importer

rmccue committed:
Check path before using it

mikeselander committed:
Merge pull request #99 from humanmade/check-error-return

Check for path errors when running via CLI

rmccue committed:
Merge pull request #87 from humanmade/fix-create-users

Fix creating users instead of mapping.

rmccue committed:
Merge pull request #79 from humanmade/already-exists-progress

Icrement progress delta when skipped / already imported items are processed

rmccue commented on class-wxr-importer.php:
```diff
@@ -1541,9 +1546,53 @@ protected function process_term( $data, $meta ) {
+			$key = apply_filters( 'import_term_meta_key', $meta_item['key'], $term_id, $term );
+			$value = false;
+
+			if ( $key ) {
```
This should be `if ( ! $key ) { continue' }`

JJJ replied:
Good catch.

rmccue commented on class-wxr-importer.php:
```diff
@@ -1541,9 +1546,53 @@ protected function process_term( $data, $meta ) {
+				return false;
+			}
+	
+			$key = apply_filters( 'import_term_meta_key', $meta_item['key'], $term_id, $term );
```
Is this an existing filter in the v1 importer? If not, should be `wxr_importer.process_term_meta.key`. Needs documentation too.

If it's not an existing filter, not sure what it offers over the above `wxr_importer.pre_process.term_meta` filter.

JJJ replied:
Agree.

stephenharris replied:
The name is adapted from https://github.com/humanmade/WordPress-Importer/blob/master/class-wxr-importer.php#L1151

In view of that, do we still want the name changed? What about post meta?

stephenharris replied:
Actually, these hooks are in the v1 importer - but they were added after this PR was opened.

rmccue commented on class-wxr-importer.php:
```diff
@@ -1541,9 +1546,53 @@ protected function process_term( $data, $meta ) {
+			 * @param array $meta_item Meta data. (Return empty to skip.)
+			 * @param int $term_id Term the meta is attached to.
+			 */
+			$meta_item = apply_filters( 'wxr_importer.pre_process.term_meta', $meta_item, $term_id );
```
Should pass `$term` here too.

JJJ replied:
Agree.

stephenharris replied:
Happy too, but I wrote this to be consistent with post meta, so we ought to change https://github.com/humanmade/WordPress-Importer/blob/master/class-wxr-importer.php#L1146 too.

(Is backwards compatibility a concern here? If not, we could just replace `$term_id` with `$term`, and similarly for posts).

rmccue commented on class-wxr-importer.php:
```diff
@@ -1541,9 +1546,53 @@ protected function process_term( $data, $meta ) {
+				}
+	
+				add_term_meta( $term_id, $key, $value );
+				do_action( 'import_term_meta', $term_id, $key, $value );
```
Is this also an existing action?

stephenharris replied:
See earlier comments :)

rmccue commented on class-wxr-importer.php:
```diff
@@ -1442,6 +1442,11 @@ protected function parse_term_node( $node, $type = 'term' ) {
 			$key = array_search( $child->tagName, $tag_name );
 			if ( $key ) {
 				$data[ $key ] = $child->textContent;
+			} else if ( $child->tagName == 'wp:termmeta' ) {
```
`else if` -> `elseif`

JJJ replied:
Agree.

rmccue commented:
@stephenharris Sorry for the long delay here :( Submitted a review; if you don't have time to fix this up, I can do so for you.

stephenharris committed:
Fix code styling.

Also removes redundant if statement.

stephenharris committed:
Fix code styling.

Also removes redundant if statement.

stephenharris committed:
Fixed code styling (for real this time)

stephenharris committed:
Merge branch 'feature/term-meta' of github.com:stephenharris/WordPress-Importer into feature/term-meta

stephenharris commented:
I clearly messed up re-basing this. I'm going to squash my commits into one and forcibly push it onto my feature branch...

stephenharris commented:
Decided to open a new PR for this: #106 

Sorry for the noise all.

rmccue commented:
In future, please don't rebase, just merge `master` into your branch regularly :)
